code/__HELPERS/path.dm

```diff
@@ -0,0 +1,379 @@
+			return
+
+		var/interesting = FALSE // have we found a forced neighbor that would make us add this turf to the open list?
+
+		switch(heading)
+			if(NORTH)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, WEST, NORTHWEST) || STEP_NOT_HERE_BUT_THERE(current_turf, EAST, NORTHEAST))
+					interesting = TRUE
+			if(SOUTH)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, WEST, SOUTHWEST) || STEP_NOT_HERE_BUT_THERE(current_turf, EAST, SOUTHEAST))
+					interesting = TRUE
+			if(EAST)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, NORTH, NORTHEAST) || STEP_NOT_HERE_BUT_THERE(current_turf, SOUTH, SOUTHEAST))
+					interesting = TRUE
+			if(WEST)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, NORTH, NORTHWEST) || STEP_NOT_HERE_BUT_THERE(current_turf, SOUTH, SOUTHWEST))
+					interesting = TRUE
+
+		if(interesting)
+			var/datum/jps_node/newnode = new(current_turf, unwind_node, steps_taken)
+			openc[current_turf] = newnode
+			open.Insert(newnode)
+			return
+
+/// For performing a scan in a given diagonal direction
+/datum/pathfind/proc/diag_scan_spec(turf/original_turf, heading)
+	var/steps_taken = 0
+	var/datum/jps_node/unwind_node = openc[original_turf]
+	var/turf/current_turf = original_turf
+	var/turf/lag_turf = original_turf
+
+	while(!unwind_node)
+		var/turf/older = sources[original_turf]
+		if(!older)
+			CRASH("JPS error: Diagonal scan couldn't find a home node")
+		unwind_node = openc[older]
+
+	while(TRUE)
+		if(path) // lazy way to force out when done, do better
+			return
+		lag_turf = current_turf
+		current_turf = get_step(current_turf, heading)
+		steps_taken++
+		if(!CAN_STEP(lag_turf, current_turf))
+			return
+
+		var/closeenough
+		if(mintargetdist)
+			closeenough = (get_dist(current_turf, end) <= mintargetdist)
+		if(current_turf == end || closeenough)
+			var/datum/jps_node/final_node = new(current_turf,unwind_node, steps_taken)
+			sources[current_turf] = original_turf
+			unwind_path(final_node)
+			return
+		else if(sources[current_turf]) // already visited, essentially in the closed list
+			return
+		else
+			sources[current_turf] = original_turf
+
+		if(unwind_node.number_tiles + steps_taken > maxnodedepth)
+			return
+
+		var/interesting = FALSE // have we found a forced neighbor that would make us add this turf to the open list?
+
+		switch(heading)
+			if(NORTHWEST)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, EAST, NORTHEAST) || STEP_NOT_HERE_BUT_THERE(current_turf, SOUTH, SOUTHWEST))
+					interesting = TRUE
+				else
+					lateral_scan_spec(current_turf, WEST)
+					lateral_scan_spec(current_turf, NORTH)
+			if(NORTHEAST)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, WEST, NORTHWEST) || STEP_NOT_HERE_BUT_THERE(current_turf, SOUTH, SOUTHEAST))
+					interesting = TRUE
+				else
+					lateral_scan_spec(current_turf, EAST)
+					lateral_scan_spec(current_turf, NORTH)
+			if(SOUTHWEST)
+				if(STEP_NOT_HERE_BUT_THERE(current_turf, WEST, SOUTHEAST) || STEP_NOT_HERE_BUT_THERE(current_turf, NORTH, NORTHWEST))
```

Comment:
```suggestion
				if(STEP_NOT_HERE_BUT_THERE(current_turf, EAST, SOUTHEAST) || STEP_NOT_HERE_BUT_THERE(current_turf, NORTH, NORTHWEST))
```
Otherwise, the rule of the algorithm is violated